nomic/aws/sagemaker.py

```diff
@@ -0,0 +1,197 @@
+    inputs[1].set_data_from_numpy(attention_mask, binary_data=True)
+
+    # have to set to binary since http doesn't natively support fp16
+    outputs.append(aiohttpclient.InferRequestedOutput("embedding", binary_data=True))
+    request, header_length = InferenceServerClient.generate_request_body(
+        inputs=inputs, outputs=outputs
+    )
+    return request, header_length
+
+
+def batch_sagemaker_requests(texts: List[str], batch_size=32):
+    """Yield sagemaker triton requests from specified size from list of texts.
+    One request will be yielded for each batch.
+    Batch size should be set based on GPU provisioned for sagemaker endpoint.
+
+    Args:
+        texts: List of texts to be batched.
+        batch_size: Size of each batch.
+
+    Yields:
+        Batches of data with size equal to batch_size.
+    """
+
+    tokenizer = load_tokenizer(NomicTextEmbeddingModel.nomic_embed_text_v1_5)
+    for i in range(0, len(texts), batch_size):
+        yield create_sagemaker_request_for_batch(texts[i : i + batch_size], tokenizer)
+
+
+def parse_sagemaker_response(response):
+    """
+    Parse response from a sagemaker triton server and return the embedding as a numpy array.
+
+    Args:
+        response: The response from the sagemaker triton server.
+
+    Returns:
+        Numpy array of embeddings.
+    """
+    # Parse json header size length from the response
+    header_length_prefix = (
+        "application/vnd.sagemaker-triton.binary+json;json-header-size="
+    )
+    header_length_str = response["ContentType"][len(header_length_prefix) :]
+
+    # Read response body
+    result = InferenceServerClient.parse_response_body(
+        response["Body"].read(), header_length=int(header_length_str)
+    )
+    return result.as_numpy("embedding")
+
+
+def embed_text(
+    texts: List[str], sagemaker_endpoint: str, region_name: str, batch_size=32
+):
+    """
+    Embed a list of texts using a sagemaker model endpoint.
+
+    Args:
+        texts: List of texts to be embedded.
+        sagemaker_endpoint: The sagemaker endpoint to use.
+        region_name: AWS region sagemaker endpoint is in.
+        batch_size: Size of each batch.
+
+    Returns:
+        Numpy array of embeddings.
+    """
+    if len(texts) == 0:
+        logger.warning("No texts to embed.")
+        return None
+
+    client = boto3.client("sagemaker-runtime", region_name=region_name)
+    embeddings = []
+
+    embeddings = []
+    for embed_request, header_length in batch_sagemaker_requests(
+        texts, batch_size=batch_size
+    ):
+        response = client.invoke_endpoint(
+            EndpointName=sagemaker_endpoint,
+            Body=embed_request,
+            ContentType="application/vnd.sagemaker-triton.binary+json;json-header-size={}".format(
+                header_length
+            ),
+        )
+        embeddings.append(parse_sagemaker_response(response))
+    return np.vstack(embeddings)
```

Comment:
it might be good to document somewhere that this returns `np.float16` dtyp'd arrays

Reply:
added in docstring but i'll also specify in the tutorial notebook